src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -185,26 +171,34 @@ public InternalEngine(EngineConfig engineConfig, boolean skipInitialTranslogReco
         logger.trace("created new InternalEngine");
     }
 
+    private Translog openTranslog(EngineConfig engineConfig, IndexWriter writer, boolean createNew) throws IOException {
+        final Translog.TranslogGeneration generation = loadTranslogIdFromCommit(writer);
+        final TranslogConfig translogConfig = engineConfig.getTranslogConfig();
+        if (createNew == false) {
+            translogConfig.setTranslogGeneration(generation);
+            if (generation != null && generation.translogUUID == null) {
+                // only upgrade on pre-2.0 indices...
+                Translog.upgradeLegacyTranslog(logger, translogConfig);
+            }
+        }
+        final Translog translog = new Translog(translogConfig);
+        if (generation == null) {
+            if (createNew) {
```

Comment:
Shouldn't this be `if (createNew == false)` instead?  I.e., if createNew is false we expect a translog to already be present (it's not a new index)?  Maybe move this up above inside the `if (createNew == false)`?  Maybe I'm completely confused about what createNew means... maybe it should be named isNewShard or something?